src/main/java/com/auth0/utils/tokens/SignatureVerifier.java

```diff
@@ -0,0 +1,83 @@
+package com.auth0.utils.tokens;
+
+import com.auth0.exception.IdTokenValidationException;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.exceptions.AlgorithmMismatchException;
+import com.auth0.jwt.exceptions.JWTDecodeException;
+import com.auth0.jwt.exceptions.JWTVerificationException;
+import com.auth0.jwt.exceptions.SignatureVerificationException;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.auth0.jwt.interfaces.JWTVerifier;
+import com.auth0.utils.Asserts;
+
+/**
+ * Represents the verification for an ID Token's signature used when validating an ID token.
+ */
+public abstract class SignatureVerifier {
+
+    private final JWTVerifier verifier;
+
+    /**
+     * Get a {@code SignatureVerifier} for use when validating an ID token signed using the HS256 signing algorithm.
+     *
+     * @param secret the client's secret to use when validating the token's signature.
+     * @return a {@code SignatureVerifier} for use with tokens signed using the HS256 signing algorithm.
+     */
+    public static SignatureVerifier forHS256(String secret) {
+        return new HS256SignatureVerifier(secret);
+    }
+
+    /**
+     * Get a {@code SignatureVerifier} for use when validating an ID token signed using the RS256 signing algorithm.
+     * Callers should provide an implementation of the {@linkplain PublicKeyProvider} to provide the public key used
+     * to verify the ID token's signature.
+     *
+     * @see PublicKeyProvider
+     * @param publicKeyProvider an implementation of {@linkplain PublicKeyProvider} to get the public key.
+     * @return a {@code SignatureVerifier} for use with tokens signed using the RS256 signing algorithm.
+     */
+    public static SignatureVerifier forRS256(PublicKeyProvider publicKeyProvider) {
+        return new RS256SignatureVerifier(publicKeyProvider);
+    }
+
+    /**
+     * Creates a new JWT Signature Verifier. Used by internal implementations to create concrete verifiers.
+     *
+     * @param verifier the instance that knows how to verify the signature. Must not be {@code null}.
+     */
+    SignatureVerifier(JWTVerifier verifier) {
+        Asserts.assertNotNull(verifier, "verifier");
+        this.verifier = verifier;
+    }
+
+    /**
+     * Verifies a token's signature.
+     *
+     * @param token the token for which to verify its signature.
+     * @return a {@linkplain DecodedJWT} that represents the token.
+     * @throws IdTokenValidationException if the signature verification failed.
+     */
+    DecodedJWT verifySignature(String token) throws IdTokenValidationException {
+        DecodedJWT decoded = decodeToken(token);
+
+        try {
+            verifier.verify(decoded);
+        } catch (AlgorithmMismatchException algorithmMismatchException) {
+            throw new IdTokenValidationException("Token signed with an unexpected algorithm", algorithmMismatchException);
```

Comment:
Ideally, it should be
```
Signature algorithm of "${decoded.header.alg}" is not supported. Expected the ID token to be signed with "${this.algorithm}"
```
How can we improve a bit the message, with the information we have available, in order to explain the expectations to the dev?

Reply:
I think we could implement a field or static getter on the verifiers that returns the algorithm name, and use that. I'll take a look.

Reply:
I've pushed a commit to expose a `getAlgorithm` method on the `SignatureVerifier`, and use it along with the algorithm of the actual `DecodedJWT` to improve the error messaging as suggested.

Reply:
Per discussion, I've pushed a commit that refactors `SignatureVerifier` to take an `Algorithm` instead of a `JWTVerifier`. This way we can avoid duplicating some of the logic to the verifier (such as ignoring the issued at claim), as well as use the `getName()` method of `Algorithm` when constructing error messages when the token's signature doesn't match the expected algorithm.